contract/src/lib.rs

```diff
@@ -540,6 +565,136 @@ impl BtcLightClient {
+        block_header: &Header,
+        prev_block_header: &ExtendedHeader,
+    ) {
+        let config = self.get_config();
+
+        if let Some(pow_allow_min_difficulty_blocks_after_height) =
+            config.pow_allow_min_difficulty_blocks_after_height
+        {
+            // Special difficulty rule for testnet:
+            // If the new block's timestamp is more than 6 * block interval minutes
+            // then allow mining of a min-difficulty block.
+            if prev_block_header.block_height >= pow_allow_min_difficulty_blocks_after_height {
+                if i64::from(block_header.time)
+                    > i64::from(prev_block_header.block_header.time)
+                        + config.pow_target_spacing(prev_block_header.block_height + 1) * 6
+                {
+                    let expected_bits = config.proof_of_work_limit_bits;
+                    require!(
+                        expected_bits == block_header.bits,
+                        format!(
+                            "Error: Incorrect target. Expected bits: {:?}, Actual bits: {:?}",
+                            expected_bits, block_header.bits
+                        )
+                    );
+                }
+            }
+        }
+
+        // Find the first block in the averaging interval
+        // and the median time past for the first and last blocks in the interval
+        let mut current_header = prev_block_header.clone();
+        const MEDIAN_TIME_SPAN: usize = 11;
+        let mut total_target = U256::ZERO;
+        let mut median_time = [0u32; MEDIAN_TIME_SPAN];
+
+        let prev_block_median_time_past = {
+            for i in 0..usize::try_from(config.pow_averaging_window).unwrap() {
+                if i < MEDIAN_TIME_SPAN {
+                    median_time[i] = current_header.block_header.time;
+                }
+
+                let (sum, overflow) = total_target
+                    .overflowing_add(target_from_bits(current_header.block_header.bits));
+                require!(!overflow, "Addition of U256 values overflowed");
+                total_target = sum;
+
+                current_header = match self
+                    .headers_pool
+                    .get(&current_header.block_header.prev_block_hash)
+                {
+                    Some(item) => item,
+                    None => return, // TODO: improve this check
+                };
+            }
+
+            median_time.sort_unstable();
+            median_time[median_time.len() / 2]
+        };
+
+        let first_block_in_interval_median_time_past = {
+            for i in 0..MEDIAN_TIME_SPAN {
+                median_time[i] = current_header.block_header.time;
+
+                current_header = match self
+                    .headers_pool
+                    .get(&current_header.block_header.prev_block_hash)
+                {
+                    Some(item) => item,
+                    None => return, // TODO: improve this check
```

Comment:
The somehow should be improved

Reply:
It seems like this should probably be a helper function since you've repeated 7 lines.